### PR TITLE
Dev remove cuda tex refs

### PR DESCRIPTION
**Pull request by griwodz**
## Description

CUDA texture references have been removed and replaced with texture objects. Instead of binding texture to memory before a kernel call and unbinding it after, the textures are now created at memory allocation time and destroyed at memory release.

The maintenance of textured CUDA memory is still global, but collected in a single object named "global_data". Untextured memory has not been touched.

Changes affect only depthMap/cuda. Prerequisite for understanding the side effects of slow kernels like refine_compUpdateYKNCCSimMapPatch_kernel .

## Implementation remarks

The changes remove global CUDA texture references, which are a limited resource and require on-demand binding to and unbinding from CUDA memory. An unlimited number of CUDA texture objects can be allocated. Using texture objects has some advantages:
 - slightly faster because bind/unbind is removed
 - texture objects are passed as parameters to CUDA kernels, which makes dependencies explicitly visible throughout the call chain
 - a piece of memory can be allocated with a texture object, keeping it throughout the memory's lifetime

CUDA memory is frequently allocated like local variables and deallocated at function exit. For memory with textures, this wasteful behavior is now avoided. Memory is now drawn from a pool in global_data, and pushed back when is not needed any more. For low end GPUs, a compile flag or explicit release may be desirable to keep the old semantics.

A surprising amount of dead code was removed in the process.

### CODE REVIEW

File: src/aliceVision/depthMap/cuda/commonStructures.hpp [hunk 1]
```diff
@@ -614,12 +671,12 @@ template<class Type, unsigned Dim> void copy(CudaDeviceMemoryPitched<Type, Dim>&
...
-    int camId;
-    int rc;
-    float* H;
-    int scale;
-    int blurid;
+    CudaHostMemoryHeap<uchar4, 2>* tex_rgba_hmh = nullptr;
+    int camId = -1;
+    int rc = -1;
+    float* H = nullptr;
+    int scale = -1;
```
[the remark this answers is not in the document]
Comment (griwodz) [reply src/aliceVision/depthMap/cuda/commonStructures.hpp]: Not correct. All members except blurid are required.

**Comment by griwodz** [comment]
I have no idea how to make Codacy happy.

**Comment by fabiencastan** [comment]
Yes, sometimes codacy detects false positives...

**Comment by fabiencastan** [comment]
I just tried the PR, but get the following errors:
```
CUDAError: misaligned address
  file:       ./src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu
  function:   ps_aggregatePathVolume
  line:       428

CUDAError: misaligned address
  file:       ./src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu
  function:   ps_aggregatePathVolume
  line:       435

CUDAError: misaligned address
  file:       ./src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu
  function:   ps_aggregatePathVolume
  line:       446

Putting textured pitched mem 378X416

CUDAError: misaligned address
  file:       ./src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu
  function:   ps_updateAggrVolume
  line:       546

CUDAError: misaligned address
  file:       ./src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu
  function:   ps_updateAggrVolume
  line:       559

CUDAError: misaligned address
  file:       ./src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu
  function:   ps_listCUDADevices
  line:       162
```

**Comment by fabiencastan** [comment]
Works fine on a 1080 card (with cuda-7.0 too).

**Comment by griwodz** [comment]
I cannot find documentation for alignment problems when the allocation uses pitched memory. The pitched mem allocation functions have no error checking, which can mean that the modified code runs out of memory.

I have added tests and exit(-1) to all the allocation functions in depthMap/cuda/commonStructures.hpp

Could you try again on the Quadro card?

If out of memory is really the problem then I need add a hook into GlobalData for a garbage collector that I had hoped to avoid.

**Comment by griwodz** [comment]
I may have found a reason.
2D textures have a maximum size of elements in horizontal and vertical direction. However, the documentation hides that they have also a byte-size limit of 4 per element in the horizontal direction, while we are using int4.
The CC 6.* cards (like the 1080) have increased the limit. That increase is undocumented, but when you ask the card itself using cudaGetDeviceProperties, it tells you.

**Comment by griwodz** [comment]
I've rebased this branch onto develop hoping that the latest changes fix the CC 3.0 problems. While it works locally, Travis stumbles over an error that I've not seen before. Can anyone help with this error?

CMake Error: install(EXPORT "aliceVision-targets" ...) includes target "aliceVision_system" more than once in the export set.

CMake seems to finish after disabling TARGETS aliceVision_system in src/aliceVision/system/CMakeLists.txt. But I don't know if that is a sensible way to fix the problem.

**Comment by griwodz** [comment]
Could someone test this pull request on one the Quadro cards with CC 3.0?
With 10 of the 3120x2080 photos from Luxembourg, depthMap results remain unchanged on a CC 5.2 and a CC 6.2 card when at least 650 MB of free CUDA memory is available.
Results are also unchanged compared the develop branch (ie. EXR files after depthMap stage are identical apart form timestamp).

**Comment by fabiencastan** [comment]
I will try that today.

**Comment by fabiencastan** [comment]
I fixed that and I'm trying the next steps.

**Comment by fabiencastan** [comment]
```
Program called with the following parameters:
 * downscale = 2
 * ini = "/tmp/MeshroomCache/PrepareDenseScene/b0111d6ecc5dc398ca0b7361cb2eb0939afc75cf/mvs.ini"
 * output = "/tmp/MeshroomCache/DepthMap/34a0bd608b0d8f568c0514332a34b8f37c0a4e39"
 * rangeSize = 3
 * rangeStart = 0
 * refineGammaC = 15.5
 * refineGammaP = 8
 * refineMaxTCams = 6
 * refineNDepthsToRefine = 31
 * refineNSamplesHalf = 150
 * refineNiters = 100
 * refineSigma = 15
 * refineUseTcOrRcPixSize = 0
 * refineWSH = 3
 * sgmGammaC = 5.5
 * sgmGammaP = 8
 * sgmMaxTCams = 10
 * sgmWSH = 4
 * verboseLevel = "info"

[12:46:05.257339][info] CUDA-Enabled GPU.
Device information:
	- id:                      0
	- name:                    Quadro K4000
	- compute capability:      3.0
	- total device memory:     3071 MB 
	- device memory available: 1994 MB 
	- per-block shared memory: 49152
	- warp size:               32
	- max threads per block:   1024
	- max threads per SM(X):   2048
	- max block sizes:         {1024,1024,64}
	- max grid sizes:          {2147483647,65535,65535}
	- number of SM(x)s:        4
	- registers per SM(x):     65536
	- registers per block:     65536
	- concurrent kernels:      yes
	- mapping host memory:     yes
	- unified addressing:      yes
	- texture alignment:       512 byte
	- pitch alignment:         32 byte

[12:46:05.317336][info] Supported CUDA-Enabled GPU detected.
[12:46:05.317799][info] Found 1 image dimension(s): 
[12:46:05.317826][info]  - [4032x3024]
[12:46:05.349897][info] Overall maximum dimension: [2016x1512]
[12:46:05.350027][info] Create depth maps.
[12:46:05.350374][info] Number of GPU devices: 1, number of CPU threads: 24
[12:46:05.350421][info] PSSGM autoScaleStep: scale: 2, step: 2
[12:46:05.350537][info] PlaneSweepingCuda:
	- max GPU Memory (MB): 1795.25
	- nImgsInGPUAtTime: 6
	- scales: 2
	- subPixel: Yes
	- varianceWSH: 4
number of CUDA devices:	1
   0: Quadro K4000
       max 1D texture size: 65536 bytes
       max 1D surface size: 65536 bytes
       max 2D texture size: (65536,65536) bytes
       max 2D surface size: (65536,32768) bytes
    INFO ps_deviceAllocate Call 1 to ps_deviceAllocate
[12:46:05.352387][info] Compute depth map: /tmp/MeshroomCache/DepthMap/34a0bd608b0d8f568c0514332a34b8f37c0a4e39/84215603_idDepthMap_scale2_step2_SGM.png
[12:46:05.352489][warning] rc: 0 - found only 4/10 tc by seeds
[12:46:05.437237][info] Camera id 0 was not in the cache, so it loads the image from disk.
CUDA device no 0 for 0
Device 0 memory - used: 1076.589844, free: 1994.722656, total: 3071.312500
Device 0 memory - used: 1165.839844, free: 1905.472656, total: 3071.312500
    INFO ps_deviceUpdateCam Calling with scales=2
Number of GPUs in the system: 1
    INFO ps_deviceUpdateCam Allocating pitch memory with 2016X1512 entries
Fetching Gaussian table for radius 2 and delta 1
Computing Gaussian table for radius 2 and delta 1
    INFO ps_deviceUpdateCam Allocating pitch memory with 1008X756 entries
[12:46:07.053163][info] Camera id 2 was not in the cache, so it loads the image from disk.
    INFO ps_deviceUpdateCam Calling with scales=2
Number of GPUs in the system: 1
    INFO ps_deviceUpdateCam Allocating pitch memory with 2016X1512 entries
Fetching Gaussian table for radius 2 and delta 1
    INFO ps_deviceUpdateCam Allocating pitch memory with 1008X756 entries
[12:46:08.079255][info] Split processing in N blocks (ntimes): 3
Number of GPUs in the system: 1
Device 0 memory - used: 1295.597656, free: 1775.714844, total: 3071.312500
total size of volume map in GPU memory: 127.722656
Number of GPUs in the system: 1
nDepths 692, nDepthsToSearch 692 
volPixs_hmh_x.getBytes(): 786432
volPixs_hmh_x.getSize().dim(): 2
volPixs_hmh_x.getSize(): (65536, 3)
nDepths 692, nDepthsToSearch 692 
Allocate textured pitched mem 65536X3
terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to allocate texture object
Failed to allocate a CUDA Texture object in /s/apps/packages/mikrosVfx/multiview/aliceVision/faca/repo/src/aliceVision/depthMap/cuda/planeSweeping/cuda_global_data.cuh 76 - invalid argument (aliceVision::depthMap::PitchedMem_Texture<T, fMode, rMode>::PitchedMem_Texture(int, int) [with T = int; cudaTextureFilterMode fMode = (cudaTextureFilterMode)0u; cudaTextureReadMode rMode = (cudaTextureReadMode)0u])
```

**Comment by griwodz** [comment]
Thanks, that error message looks better than the previous ones.
Was the "cannot get memory info" a bug in the code?

**Comment by griwodz** [comment]
More than 100 commits made it really hard to rebase onto develop again. I would therefore prefer to close this pull request and continue in [PR 460](url).

**Pull request closed without merging**